staging/src/k8s.io/endpointslice/topologies/preferzonetopology.go

```diff
@@ -0,0 +1,145 @@
+	"k8s.io/klog/v2"
+)
+
+// PreferZoneTopology is a topology that provides EndpointSlice topology hints
+// using the [PreferZone Heuristic].
+//
+// [PreferZone Heuristic]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2433-topology-aware-hints#preferzone-heuristic
+type PreferZoneTopology struct{}
+
+// PreferZoneTopology implements the Topology interface.
+var _ Topology = &PreferZoneTopology{}
+
+func NewPreferZoneTopology() *PreferZoneTopology {
+	return &PreferZoneTopology{}
+}
+
+// Name returns the name associated with the PreferZoneTopology. This will be
+// matched against the topology annotation in the Service to decide if the
+// Topology is currently active for a service.
+func (t *PreferZoneTopology) Name() string {
+	return "PreferZone"
+}
+
+// PopulateHints populates topology hints on EndpointSlices and returns
+// updated lists of EndpointSlices to create and update. It also returns any
+// Events that need to be recorded for the Service.
+//
+// Combination of [SliceInfo.ServiceKey and SliceInfo.AddressType] are used to
+// identify the endpoint slices which these hints will be associated with.
+func (t *PreferZoneTopology) PopulateHints(logger klog.Logger, si *SliceInfo) (slicesToCreate []*discoveryv1.EndpointSlice, slicesToUpdate []*discoveryv1.EndpointSlice, events []*EventBuilder) {
+	// Step 1: Ensure that all endpoints within EndpointSlices have a zone
+	// assigned.
+	if si.getEndpointsWithMissingZone() != 0 {
+		logger.Info("Endpoint found without zone specified, removing hints", "key", si.ServiceKey, "addressType", si.AddressType)
+		events = append(events, &EventBuilder{
+			EventType: v1.EventTypeWarning,
+			Reason:    "TopologyAwareHintsDisabled",
+			Message:   FormatWithAddressTypeAndTopologyName(NoZoneSpecified, si.AddressType, t.Name()),
+		})
+		slicesToCreate, slicesToUpdate := RemoveHintsFromSlices(si)
+		return slicesToCreate, slicesToUpdate, events
+	}
+
+	// Step 2: Identify the Unchanged slices that need an update because of
+	// missing or incorrect zone hint.
+	//
+	// Using filtering in place to remove any endpoints that are no longer
+	// Unchanged and need to be moved to ToUpdate
+	// (https://github.com/golang/go/wiki/SliceTricks#filter-in-place)
+	j := 0
+	for _, slice := range si.Unchanged {
+		if t.sliceNeedsUpdate(slice) {
+			// Refer the comment in SliceInfo.Unchanged to understand the need to
+			// DeepCopy.
+			si.ToUpdate = append(si.ToUpdate, slice.DeepCopy())
+		} else {
+			si.Unchanged[j] = slice
+			j++
+		}
+	}
+	// truncate si.Unchanged so it only includes slices that are still
+	// unchanged.
+	si.Unchanged = si.Unchanged[:j]
+
+	// Step 3: Add same zone hints to all slices that need to be created or
+	// updated.
+	t.addSameZoneHint(si.ToCreate)
+	t.addSameZoneHint(si.ToUpdate)
+
+	logger.Info("Topology Aware Hints has been enabled, adding hints.", "key", si.ServiceKey, "addressType", si.AddressType, "topology", t.Name())
```

Comment:
Wouldn't this log anytime EndpointSlices were reconciled? If so, we probably want a different log message here (or to be able to detect if it's the first time hints are being added).